services/cd-service/pkg/repository/transformer.go

```diff
@@ -365,8 +366,25 @@ type transformerRunner struct {
 	DeletedRootApps []RootApp
 }
 
+var isolatedTransformersLock sync.RWMutex
+var isolatedTransformerNames = []string{"UndeployApplication", "DeleteEnvFromApp", "DeleteEnvironment"}
+
 func (r *transformerRunner) Execute(ctx context.Context, t Transformer, transaction *sql.Tx) error {
 	r.Stack = append(r.Stack, nil)
+	tranformerTypeName := reflect.TypeOf(t).Name()
+	isIsolated := false
+	for _, isolatedTransformerName := range isolatedTransformerNames {
+		if isolatedTransformerName == tranformerTypeName {
+			isIsolated = true
+		}
+	}
+	if isIsolated {
+		isolatedTransformersLock.Lock()
+		defer isolatedTransformersLock.Unlock()
+	} else {
+		isolatedTransformersLock.RLock()
+		defer isolatedTransformersLock.RUnlock()
+	}
```

Comment:
I think this requires a bit of explanation:
```suggestion
	if isIsolated {
	        // we protect all "destructive" operations by a read-write lock, so that only 1 destructive operation can be run in parallel:
		isolatedTransformersLock.Lock()
		defer isolatedTransformersLock.Unlock()
	} else {
	        // we also use a read lock, so that destructive and non-destructive transformers cannot run in parallel:
		isolatedTransformersLock.RLock()
		defer isolatedTransformersLock.RUnlock()
	}
```